src/hash.cr

```diff
@@ -118,6 +290,479 @@ class Hash(K, V)
+    # We actually store 1 more than the value because 0 means empty.
+    value += 1
+
+    # We also have to see what we have: UInt8, UInt16 or UInt32 buckets.
+    case @indices_bytesize
+    when 1
+      @indices[index] = value.to_u8!
+    when 2
+      @indices.as(UInt16*)[index] = value.to_u16!
+    else
+      @indices.as(UInt32*)[index] = value.to_u32!
+    end
+  end
+
+  # Marks `@indices` at `index` as empty.
+  private def empty_index(index) : Nil
+    case @indices_bytesize
+    when 1
+      @indices[index] = 0_u8
+    when 2
+      @indices.as(UInt16*)[index] = 0_u16
+    else
+      @indices.as(UInt32*)[index] = 0_u32
+    end
+  end
+
+  # Returns the capacity of `@indices`.
+  private def indices_size
+    1 << @indices_size_pow2
+  end
+
+  # Computes what bytesize we'll store in `@indices` according to its size
+  private def compute_indices_bytesize(size) : Int8
+    case
+    when size <= MAX_INDICES_BYTESIZE_1
+      1_i8
+    when size <= MAX_INDICES_BYTESIZE_2
+      2_i8
+    else
+      4_i8
+    end
+  end
+
+  # Allocates `size` number of indices for `@indices`.
+  private def malloc_indices(size)
+    Pointer(UInt8).malloc(size * @indices_bytesize)
+  end
+
+  # Reallocates `size` number of indices for `@indices`.
+  private def realloc_indices(size)
+    @indices.realloc(size * @indices_bytesize)
+  end
+
+  # Marks all existing indices as empty.
+  private def clear_indices : Nil
+    @indices.clear(indices_size * @indices_bytesize)
+  end
+
+  # Returns the entry in `@entries` at `index`.
+  private def get_entry(index) : Entry(K, V)
+    @entries[index]
+  end
+
+  # Sets the entry in `@entries` at `index`.
+  private def set_entry(index, value) : Nil
+    @entries[index] = value
+  end
+
+  # Adds an entry at the end and also increments this hash's size.
+  private def add_entry_and_increment_size(hash, key, value) : Nil
+    set_entry(entries_size, Entry(K, V).new(hash, key, value))
+    @size += 1
+  end
+
+  # Marks an entry in `@entries` at `index` as deleted
+  # *without* modifying any counters (`@size` and `@deleted_count`).
+  private def delete_entry(index) : Nil
+    set_entry(index, Entry(K, V).deleted)
+  end
+
+  # Marks an entry in `@entries` at `index` as deleted
+  # and updates the `@size` and `@deleted_count` counters.
+  private def delete_entry_and_update_counts(index) : Nil
+    delete_entry(index)
+    @size -= 1
+    @deleted_count += 1
+  end
+
+  # Returns true if there's no place for new entries without doing a resize.
+  private def entries_full? : Bool
+    entries_size == indices_size / 2
+  end
+
+  # Yields each non-deleted Entry with its index inside `@entries`.
+  private def each_entry_with_index
+    entries_size.times do |i|
+      entry = get_entry(i)
+      yield entry, i unless entry.deleted?
+    end
+  end
+
+  # Computes the next index in `@indices`, needed when an index is not empty.
+  private def next_index(index : Int32) : Int32
+    fit_in_indices(index + 1)
```

Comment:
"linear probing" is a bad idea.
Use at least "quadratic probing". It has same good cache locality on first three probes as for linear probing.

Reply:
In Ruby they tried quadratic probing and it turned out to be a bit slower. They use linear probing. That's why I also chose that, and it's simpler.

Ref: https://github.com/ruby/ruby/blob/c94cc6d968a7241a487591a9753b171d8081d335/st.c#L869-L872

But if you want we can try it. I know nothing about quadratic probing, is it just `index, `index + 1, index + 4, index + 9, index + 16, etc.`?

Reply:
According to the ref they aren't using linear probing now, they are using Double Hashing.

Reply:
It's true! However I don't understand how that works:

     Still other bits of the hash value are used when the mapping
     results in a collision.  In this case we use a secondary hash
     value which is a result of a function of the collision bin
     index and the original hash value.  The function choice
     guarantees that we can traverse all bins and finally find the
     corresponding bin as after several iterations the function
     becomes a full cycle linear congruential generator because it
     satisfies requirements of the Hull-Dobell theorem.

I don't know about that theorem so it's harder for me to implement something I don't understand. But I'll definitely have that in mind for further optimizing this (but it'll me some more time, but PRs are welcome!).